cluster_tools/remote.py

```diff
@@ -92,26 +92,35 @@ def worker(workerid, job_array_index, cfut_dir):
     logging.info("Pickle file renamed to {}.".format(destfile))
 
 
-def setup_logging(meta_data):
-    logging_config = meta_data.get(
-        "logging_config",
-        {"level": logging.DEBUG, "format": "%(asctime)s %(levelname)s %(message)s"},
-    )
+def setup_logging(meta_data, log_file_path):
+    if "logging_setup_fn" in meta_data:
+
+        # As this is the file the cluster_tools themselves are logging to, append a suffix
+        # TODO: This should probably be done by the caller, but then what if the caller doesn't
+        # change the log file name - then sbatch and the caller would log to the file
+        meta_data["logging_setup_fn"](log_file_path + ".debug")
```

Comment:
is it possible to check which output file the caller chose after calling the setup function here? if so, one could assert that the file handle is different to `log_file_path`. however, I'm not too confident that this is possible at all.
an alternative but also but cumbersome approach could be the following:
the caller passes `logging_setup_fn` and `prepare_log_file_path`. the latter function would be responsible to adapt the path (e.g., append `.debug`). `remote.py` would call that prepare-function, assert that the name was changed and then pass the name to `logging_setup_fn`. the only downside I see is that two functions would need to be passed instead of only one..

Reply:
Hmm, yeah, an alternative would be to return the adapted log file path in the `logging_setup_fn` but that would also be a somewhat obscure requirement :/
Maybe the error case (that both attempt to log to the same file) doesn't need to be prevented in our code, since it would be noticed rather quickly when trying to open the same file from two processes with write permission (which should crash)?

In any case, there should be a spot in the code where the parameters to the executor (`logging_config`, `logging_setup_fn`, ...) are documented and where we can list the requirements and parameters to the function(s).

Reply:
> Maybe the error case (that both attempt to log to the same file) doesn't need to be prevented in our code, since it would be noticed rather quickly when trying to open the same file from two processes with write permission (which should crash)?

Very good point. That would be ideal and is probably also the case. Maybe you can test this?

> In any case, there should be a spot in the code where the parameters to the executor (logging_config, logging_setup_fn, ...) are documented and where we can list the requirements and parameters to the function(s).

Yes, probably a docstring for the method which also accepts the `logging_setup_fn` parameter, right?

Reply:
> Very good point. That would be ideal and is probably also the case. Maybe you can test this?

Unfortunately, no error is thrown :/ It seems like the python debug file logging is "winning", but at the end of the log file there is a couple of duplicated lines...

I'm hesitating to implement the `prepare_log_file_path` function, because it complicates the interface and needs to be supplied on every `get_executor` call. I'm wondering whether the added "safety" is worth that. I would suggest to document this requirement (to adapt the log file path) in the `get_executor` docstring and leave it at that for now. Would you be ok with that or prefer to have the added safety?

Reply:
I agree, that the added safety is probably not worth it. So, your suggested approach sounds fine to me :+1: However, I just had another idea in which both parties would make use of a suffix:
- CT passes `log_file_path` to the caller (which is supposed to use that with a suffix à la `"f{log_file_path}.some_suffix"`
- CT itself also adds a suffix when scheduling the job via slurm (e.g., `"f{log_file_path}.native`)

The wording might be tricky (and obviously the caller shouldn't use the same suffix), but overall the risk of walking into the pitfall would be way smaller. However, as I said before, I'm also fine with your suggestion :)

Reply:
That's a nice trade off :+1: